Snapping

RareScrap opened the pull request:
Issue #14 

+ Extracted defaults

RareScrap committed:
appbar snapping implementation for ExitUntilCollapsed scroll strategy

+ Extracted defaults

RareScrap committed:
appbar snap implementation

onebone commented:
@RareScrap I have scanned your code so far, I have some possible options that can deal with your situation.

For scroll strategies other than `ExitUntilCollapsed`, they need to modify `offsetY` which is a member of `CollapsingToolbarScaffoldState`. The cause derives from the fact that snap behavior is actually a feature of `CollapsingToolbarScaffold`. I expect the problem would be fixed if you move the snapping methods to ScaffoldState.


And for interrupting animation with user inputs, I think it may be solved by utilizing `MutatePriority` though it may need some extra researches.

RareScrap committed:
moved snap processing from CollapsingToolbarState to CollapsingToolbarScaffoldState

+ Defaults naming refactoring
+ TODOs added

RareScrap commented:
Thank you for your review. I was little busy last weekend. Will done with PR at this weekend.

onebone commented:
Great!

RareScrap commented:
There are many TODOs that require your attension. Please comment on each of them so we can discuss them.

onebone commented:
Thank you for reminding, I will review it until this weekend!

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
 
 	internal val offsetYState = mutableStateOf(initialOffsetY)
+
+	// TODO: Maybe should move toolbar expand/collapse methods to CollapsingToolbarState
+	//  but offset expand/collapse leave in CollapsingToolbarScaffoldState?
+
+	// TODO: A strange jump in snap speed is often observed
+	@ExperimentalToolbarApi
+	suspend fun expand(duration: Int = CollapsingToolbarDefaults.ExpandDuration) {
+		val anim = AnimationState(toolbarState.height.toFloat())
+
+		toolbarState.scroll {
+			var prev = anim.value
+			anim.animateTo(toolbarState.maxHeight.toFloat(), tween(duration)) {
+				scrollBy(value - prev)
+				prev = value
+			}
+		}
+	}
+
+	// TODO: A strange jump in snap speed is often observed
+	@ExperimentalToolbarApi
+	suspend fun collapse(duration: Int = CollapsingToolbarDefaults.CollapseDuration) {
+		val anim = AnimationState(toolbarState.height.toFloat())
+
+		toolbarState.scroll {
+			var prev = anim.value
+			anim.animateTo(toolbarState.minHeight.toFloat(), tween(duration)) {
+				scrollBy(value - prev)
+				prev = value
+			}
+		}
+	}
```
`expand` and `collapse` method does not do anything with a property of `CollapsingToolbarScaffoldState`, it would be more appropriate to leave it in `CollapsingToolbarState`.

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
+		}
+	}
+
+	// TODO: Is there a better solution rather OptIn ExperimentalToolbarApi?
```
What makes it bother you??

RareScrap replied:
I can't decide what to use: `@ExperimentalToolbarApi` or `@OptIn(ExperimentalToolbarApi::class)`. I suppose that `@ExperimentalToolbarApi` is using for public animation methods and `@OptIn(ExperimentalToolbarApi::class)` safe to use in internal methods but I'm not sure.

What are you think? Why you using `@ExperimentalToolbarApi` at all?

RareScrap replied:
> Expierence

Sorry, I don't understand you. What exactly do you mean by this?

onebone commented on lib/src/main/java/me/onebone/toolbar/SnapStrategy.kt:
```diff
@@ -0,0 +1,11 @@
+	@FloatRange(from = 0.0, to = 1.0) val edge: Float = CollapsingToolbarDefaults.Edge,
+	val expandDuration: Int = CollapsingToolbarDefaults.ExpandDuration,
+	val collapseDuration: Int = CollapsingToolbarDefaults.CollapseDuration
+)
```
Please place a new line at the end of the file

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -75,14 +151,16 @@ fun CollapsingToolbarScaffold(
 
 	val nestedScrollConnection = remember(scrollStrategy, state) {
-		scrollStrategy.create(state.offsetYState, state.toolbarState, flingBehavior)
+		// TODO by RareScrap: Should we make offsetYState public and pass just state?
```
Thank you for pointing this, I think having reference to the same `MutableState<T>` object in many different sites is error-prone and should be removed. Now you have added `expandOffset` and `collapseOffset`, they might be removed in favor of these methods.

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
+	}
+
+	// TODO: Is there a better solution rather OptIn ExperimentalToolbarApi?
+	@OptIn(ExperimentalToolbarApi::class)
+	internal suspend fun processSnap(strategy: SnapStrategy) {
+		if (toolbarState.progress > strategy.edge) {
+			expand(strategy.expandDuration)
+		} else {
+			collapse(strategy.collapseDuration)
+		}
+	}
```
I see no reason for `processSnap` to reside in ScaffoldState. You can put it as an extension function like this:

```kotlin
@OptIn(ExperimentalToolbarApi::class)
internal suspend fun SnapStrategy.performTo(state: CollapsingToolbarState) {
	if (state.progress > edge) {
		state.expand(expandDuration)
	} else {
		state.collapse(collapseDuration)
	}
}
```

onebone replied:
It makes sense if the context parameter is `CollapsingToolbarState` as well.

```kotlin
@OptIn(ExperimentalToolbarApi::class)
internal suspend fun CollapsingToolbarState.performSnap(strategy: SnapStrategy) {
	if (state.progress > edge) {
		expand(strategy.expandDuration)
	} else {
		collapse(strategy.collapseDuration)
	}
}
```

RareScrap replied:
Sounds good but I need to research would extensions be slower than regular method. Also I don't know where to place this extensions. I think they should be private and store inside `SnapStrategy` file.

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
+		}
+	}
+
+	@ExperimentalToolbarApi
+	suspend fun expandOffset(snapStrategy: SnapStrategy) {
+		val anim = AnimationState(offsetYState.value.toFloat())
+
+		anim.animateTo(0f, tween(snapStrategy.expandDuration)) {
+			offsetYState.value = value.toInt()
+		}
+	}
+
+	@ExperimentalToolbarApi
+	suspend fun collapseOffset(snapStrategy: SnapStrategy) {
+		val anim = AnimationState(offsetYState.value.toFloat())
+
+		anim.animateTo(-toolbarState.minHeight.toFloat(), tween(snapStrategy.collapseDuration)) {
+			offsetYState.value = value.toInt()
+		}
+	}
```
I think it would be more flexible to use if its parameter is duration instead. `CollapsingToolbarScaffoldState` seems to have no need to know anything about `SnapStrategy`.

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
+		}
+	}
+
+	// TODO: Is there a better solution rather OptIn ExperimentalToolbarApi?
+	@OptIn(ExperimentalToolbarApi::class)
+	internal suspend fun processOffsetSnap(snapStrategy: SnapStrategy) {
+		// TODO: Refactor ugly math
+		val offsetProgress =
+			1f - ((offsetYState.value / (toolbarState.minHeight / 100f)) / 100f).absoluteValue
+		if (offsetProgress > snapStrategy.edge) {
+			expandOffset(snapStrategy)
+		} else {
+			collapseOffset(snapStrategy)
+		}
+	}
```
This method can also be refactored as an extension function!

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
+	internal suspend fun processOffsetSnap(snapStrategy: SnapStrategy) {
+		// TODO: Refactor ugly math
+		val offsetProgress =
+			1f - ((offsetYState.value / (toolbarState.minHeight / 100f)) / 100f).absoluteValue
```
This expression seems to be identical to `1f - (offsetY / toolbarState.minHeight).absoluteValue`.

We might have to add a branch to check if `minHeight` is zero.

ChristopherKlammt commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbarScaffold.kt:
```diff
@@ -45,6 +49,78 @@ class CollapsingToolbarScaffoldState(
+	}
+
+	// TODO: Is there a better solution rather OptIn ExperimentalToolbarApi?
+	@OptIn(ExperimentalToolbarApi::class)
+	internal suspend fun processSnap(strategy: SnapStrategy) {
+		if (toolbarState.progress > strategy.edge) {
+			expand(strategy.expandDuration)
+		} else {
+			collapse(strategy.collapseDuration)
+		}
+	}
```
Should add a check for whether the toolbar needs to be snapped at all.. otherwise running into some weird visual bugs on horizontal scrolls inside the CollapsingToolbarScaffold content.

```suggestion
	@OptIn(ExperimentalToolbarApi::class)
	internal suspend fun processSnap(strategy: SnapStrategy) {
           if (progress != 1f && progress != 0f) {
              if (progress > strategy.edge) {
                expand(strategy.expandDuration)
            } else {
                collapse(strategy.collapseDuration)
            }
        }
	}
```

RareScrap replied:
Wonderful idea. Thank you for sharing. I think it worth to refactor `progress != 1f && progress != 0f` to remove unnecessary if-nesting. Can you do it please?

RareScrap replied:
Also you can use Kotlin range syntax for it:
```kotlin
if (progress in 0f..1f)
```

ChristopherKlammt replied:
Created pull-request, please have a look at it here: https://github.com/RareScrap/compose-collapsing-toolbar/pull/1

RareScrap replied:
Merged. Thank you for your participation.

RareScrap committed:
added end line

RareScrap committed:
moved toolbar expand/collapse methods to CollapsingToolbarState

RareScrap committed:
removed redundant offsetY passing

RareScrap committed:
processSnap and processOffsetSnap moved into extensions

RareScrap committed:
refactored argument of collapseOffset and expandOffset

RareScrap committed:
refactored ugly math

RareScrap committed:
renamed SnapStrategy to SnapConfig

ChrisKlammt committed:
Add check whether snap is needed

- when toolbar is expanded or collapsed fully, do not perform snap (as it is not needed), this is the case, when progress == 0 or 1
- this improves performance and fixes a bug with horizontal scrolling in the CollapsingToolbarScaffold content

RareScrap committed:
Merge pull request #1 from ChristopherKlammt/feature/snap

Add check whether snap to perform snap

ChristopherKlammt commented:
@onebone what are the plans for merging this PR?

RareScrap commented:
There is a bug in it. I will describe it tomorrow.

Marx6 reviewed (commented):
Coded

onebone commented on lib/src/main/java/me/onebone/toolbar/ScrollStrategy.kt:
```diff
@@ -246,6 +282,32 @@ internal class ExitUntilCollapsedNestedScrollConnection(
+}
+
+// TODO: Is there a better solution rather OptIn ExperimentalToolbarApi?
+@OptIn(ExperimentalToolbarApi::class)
```
`expandOffset` and `collapseOffset` are experimental because it is subject to change their behavior or surface anytime. Meantime, `performSnap` is private and we don't have to be warned by this potential change so I think it would be sufficient.

onebone commented on lib/src/main/java/me/onebone/toolbar/AppBarContainer.kt:
```diff
@@ -76,15 +67,15 @@ fun AppBarContainer(
-	collapsingToolbarState: CollapsingToolbarState,
+	collapsingToolbarScaffoldState: CollapsingToolbarScaffoldState,
 	content: @Composable AppbarContainerScope.() -> Unit
 ) {
```
Please revert contract(i.e. parameter or behavior) changes in `AppBarContainer.kt` as it is in maintenance mode and thus should not be changed until it is removed!

onebone commented on lib/src/main/java/me/onebone/toolbar/CollapsingToolbar.kt:
```diff
@@ -136,8 +125,23 @@ class CollapsingToolbarState(
+		return left
+	}
+
+	// TODO: A strange jump in snap speed is often observed
```
Can you explain this issue?

RareScrap replied:
This is a bug I mentioned. You can see a strange expanding before the snap will applied. This sometimes works even in wrong direction. Look at the demonstration GIF:
<img src="https://user-images.githubusercontent.com/7997602/150323590-52614361-5e96-4436-a595-c552ebcb6b40.gif" width="300px">

When I release the finger you can see that snapping is moving in wrong direction (to top) before it finally snap to full size.

I tried to debug it and I suppose that a bug is in implementation of `EnterAlwaysCollapsedNestedScrollConnection#onPostScroll()` method. Unfortunately I don't know how `NestedScrollConnection` and math in `EnterAlwaysCollapsedNestedScrollConnection` work. I will be glad if you help me with it. Try to build an example app and reproduce this bug.

RareScrap commented:
Sorry guys, I was too busy to done with it but now I have some time :)

jkwiecien commented:
Rooting for that snapping guys! :)

GIGAMOLE commented:
Hi, I created a newer version of the snapping feature here: https://github.com/onebone/compose-collapsing-toolbar/pull/83

PR contains the recorded gifs and you can test it by yourself. For me, it seems like everything is working fine.
